plugins/flow-visibility/clickhouse-monitor/main.go

```diff
@@ -0,0 +1,256 @@
+func getPodLogs() (string, error) {
+	var logString string
+	podLogOpts := corev1.PodLogOptions{}
+	config, err := rest.InClusterConfig()
+	listOptions := metav1.ListOptions{
+		LabelSelector: monitorLabel,
+	}
+	if err != nil {
+		return logString, fmt.Errorf("error when getting config: %v", err)
+	}
+	// creates the clientset
+	clientset, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		return logString, fmt.Errorf("error when getting access to K8S: %v", err)
+	}
+	// gets Clickhouse monitor pod
+	pods, err := clientset.CoreV1().Pods(namespace).List(context.TODO(), listOptions)
+	if err != nil {
+		return logString, fmt.Errorf("failed to list clickhouse monitor Pods: %v", err)
+	}
+	for _, pod := range pods.Items {
+		// reads logs from the last successful pod
+		if pod.Status.Phase == corev1.PodSucceeded {
+			req := clientset.CoreV1().Pods(namespace).GetLogs(pod.Name, &podLogOpts)
+			podLogs, err := req.Stream(context.TODO())
+			if err != nil {
+				return logString, fmt.Errorf("error when opening stream: %v", err)
+			}
+			defer podLogs.Close()
+
+			buf := new(bytes.Buffer)
+			_, err = io.Copy(buf, podLogs)
+			if err != nil {
+				return logString, fmt.Errorf("error when copying information from podLogs to buf: %v", err)
+			}
+			logString := buf.String()
+			return logString, nil
+		}
+	}
+	return logString, fmt.Errorf("no successful monitor")
+}
+
+// Connects to Clickhouse in a loop
+func connectLoop() (*sql.DB, error) {
+	// Clickhouse configuration
+	userName := os.Getenv("CH_USERNAME")
+	password := os.Getenv("CH_PASSWORD")
+	host, port := os.Getenv("SVC_HOST"), os.Getenv("SVC_PORT")
+
+	ticker := time.NewTicker(connectionWait)
+	defer ticker.Stop()
+
+	timeoutExceeded := time.After(connectionTimeout)
+	for {
+		select {
+		case <-timeoutExceeded:
+			return nil, fmt.Errorf("failed to connect to Clickhouse after %s", connectionTimeout)
+
+		case <-ticker.C:
+			// Open the database and ping it
+			dataSourceName := fmt.Sprintf("tcp://%s:%s?debug=true&username=%s&password=%s", host, port, userName, password)
+			connect, err := sql.Open("clickhouse", dataSourceName)
+			if err != nil {
+				klog.ErrorS(err, "Failed to connect to Clickhouse")
+			}
+			if err := connect.Ping(); err != nil {
+				if exception, ok := err.(*clickhouse.Exception); ok {
+					klog.ErrorS(nil, "Failed to ping Clickhouse", "message", exception.Message)
+				} else {
+					klog.ErrorS(err, "Failed to ping Clickhouse")
+				}
+			} else {
+				return connect, nil
+			}
+		}
+	}
+}
+
+// Checks the memory usage in the Clickhouse, deletes records when it exceeds the threshold.
+func monitorMemory(connect *sql.DB) bool {
+	var (
+		freeSpace  uint64
+		totalSpace uint64
+	)
+	// Get memory usage from Clickhouse system table
+	if err := connect.QueryRow("SELECT free_space, total_space FROM system.disks").Scan(&freeSpace, &totalSpace); err != nil {
+		klog.ErrorS(err, "Failed to get memory usage for Clickhouse")
+		return false
```

Comment:
do you think we should wait for the next execution of the cronjob or could instead retry locally in case of transient errors? Could it be possible to retry at least in case of timeout?

Reply:
Added a retry logic for all queries in the monitor.

Reply:
cool